lib/rouge/lexers/sas.rb

```diff
@@ -0,0 +1,568 @@
+        # case 2 - where it follows the previous statement on the line (after a semicolon)
+        rule /(;)(\s*)(%?\*[^;]*;)/m do
+          groups Punctuation, Text, Comment::Single
+        end
+
+        # True multiline comments!
+        rule %r(/[*].*?[*]/)m, Comment::Multiline
+
+        # date/time constants (Language Reference pp91-2)
+        rule /'[0-9a-z]+?'d/i, Literal::Date
+        rule /'.+?'dt/i, Literal::Date
+        rule /'[0-9:]+?([a|p]m)?'t/i, Literal::Date
+
+        rule /'/, Str::Single, :single_string
+        rule /"/, Str::Double, :double_string
+        rule /&[a-z0-9_&.]+/i, Name::Variable
+
+        # numeric constants (Language Reference p91)
+        rule /\d[0-9a-f]*x/i, Num::Hex
+        rule /\d[0-9e\-.]+/i, Num # scientific notation
+
+        # auto variables from DATA step (Language Reference p46, p37)
+        rule /\b(_n_|_error_|_file_|_infile_|_msg_|_iorc_|_cmd_)\b/i, Name::Builtin::Pseudo
+
+        # auto variable list names
+        rule /\b(_character_|_numeric_|_all_)\b/i, Name::Builtin
+
+        # datalines/cards etc
+        rule /\b(datalines|cards)(\s*)(;)/i do
+          groups Keyword, Text, Punctuation
+          push :datalines
+        end
+        rule /\b(datalines4|cards4)(\s*)(;)/i do
+          groups Keyword, Text, Punctuation
+          push :datalines4
+        end
+
+
+        # operators (Language Reference p96)
+        rule %r(\*\*|[\*/\+-]), Operator
+        rule /[^¬~]?=:?|[<>]=?:?/, Operator
+        rule /\b(eq|ne|gt|lt|ge|le|in)\b/i, Operator::Word
+        rule /[&|!¦¬∘~]/, Operator
+        rule /\b(and|or|not)\b/i, Operator::Word
+        rule /[<>|><]/, Operator # min/max
+        rule /\|\|/, Operator # concatenation
+
+        # The OF operator should also be highlighted (Language Reference p49)
+        rule /\b(of)\b/i, Operator::Word
+        rule /\b(like)\b/i, Operator::Word # Language Ref p181
+      
+        rule /\d+/, Num::Integer
+
+        rule /\$/, Keyword::Type
+
+        # Macro definitions
+        rule /(%macro|%mend)(\s*)([\w\d]+)/i do
+          groups Keyword, Text, Name::Function
+        end
+        rule /%mend/, Keyword
+
+        rule /%\w[\w\d]*/ do |m|
+          if self.class.sas_macro_statements.include? m[0].upcase
+            token Keyword
+          elsif self.class.sas_macro_functions.include? m[0].upcase
+            token Keyword
+          else
+            token Name
+          end
+        end
+      end
+
+      state :basics2 do
+        # Rules to be parsed after the keywords (which are different depending
+        # on the context)
+
+        # Missing values (Language Reference p81)
+        rule /\s\.[;\s]/, Keyword::Constant # missing
+        rule /\s\.[a-z_]/, Name::Constant # user-defined missing
+
+        rule /[\(\),;:\{\}\[\]\\\.]/, Punctuation
+
+        rule /@/, Str::Symbol # line hold specifiers
+        rule /\?/, Str::Symbol # used for format modifiers
+        
```

Comment:
This will mean errors in the syntax won't be highlighted. Is that intentional?

Reply:
Even worse, if there's an error, we this will simply render the entire rest of the stream as text...

Reply:
(actually the rest of the line - `.` doesn't match newlines since there's no `/m`)

Reply:
I of course meant to attach this comment to line 473. @jneen's comments are about the same line. Sorry for the confusion @tomsutch!

Reply:
Thanks both. I have changed to match to the end of the word, rather than to the end of the line, in https://github.com/rouge-ruby/rouge/pull/1107/commits/004c8453247957e0b7dd4d5326187eec340b7d95, which I hope is less concerning.

I think being able to reliably pick up errors is beyond me - there's such a tangle of things that are potentially valid that I'd prefer it just to fall back to rendering as text.